lib/datadog/profiling/ext/dir_monkey_patches.rb

```diff
@@ -0,0 +1,223 @@
+                yield entry_name
+                # <-- We'll go back to the Dir internals, critical region again
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              end
+            ensure
+              # <-- End critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+            end
+          else
+            # This returns an enumerator. We don't want/need to intercede here, the enumerator will eventually call the
+            # other branch once it gets going.
+            super
+          end
+        end
+        ruby2_keywords :foreach if respond_to?(:ruby2_keywords, true)
+
+        def glob(*args, &block)
+          if block
+            begin
+              # <-- Begin critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              super do |entry_name|
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+                # <-- We're safe now while running customer code
+                yield entry_name
+                # <-- We'll go back to the Dir internals, critical region again
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              end
+            ensure
+              # <-- End critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+            end
+          else
+            begin
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              super
+            ensure
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+            end
+          end
+        end
+        ruby2_keywords :glob if respond_to?(:ruby2_keywords, true)
+
+        def home(*args, &block)
+          Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+          super
+        ensure
+          Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+        end
+        ruby2_keywords :home if respond_to?(:ruby2_keywords, true)
+      end
+
+      # Monkey patches for Dir. See DirMonkeyPatches above for more details.
+      module DirInstanceMonkeyPatches
+        def each(*args, &block)
+          if block
+            begin
+              # <-- Begin critical region
+              Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
+              super do |entry_name|
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_resume_interruptions
+                # <-- We're safe now while running customer code
+                yield entry_name
+                # <-- We'll go back to the Dir internals, critical region again
+                Datadog::Profiling::Collectors::CpuAndWallTimeWorker._native_hold_interruptions
```

Comment:
:thinking: I don't think so...? It's ok to deliver signals during exception handling, so if an exception gets raised in the client code, it should be fine for the profiler to continue operating from them on, since no more directory calls will be made by Ruby.

Or do you mean making sure that `_native_hold_interruptions` never raises or something similar?

Reply:
I really don't know, hence why I ask ;)

I'm basing myself on the following outline:

```
begin
  # [A+] enter our code: hold
  super do
    # [B-] exit our code back to not-our-code: resume
    yield
    # [B+] enter our code: hold
  end
ensure
  # [A-] exit our code: resume
end
```

On the super happy path the execution order is `[A+] [B-] [B+] [A-]`.

But if `yield` raises it becomes `[A+] [B-] [A-]`, which in my mind triggers some "consistency" alarms and makes me wonder about internal details:

- what happens when `_native_resume_signals` is called twice in a row? is it a NOOP? is it doing work twice (e.g invoking a syscall) but idempotent?
- there seems to be work to be protected after `yield` (because of that `[B+]`) and before method exit (`[A-]`) so I presume that work is in `super`, but if `yield` raises there may be work to protect too in a hypothetical `ensure` block inside that `super` too? Or is `[B+]` entirely superfluous even in the happy `yield` case?

Conversely, imagine an abstraction like so:

```
module SignalHolder
  def hold
	# hold
    yield
  ensure
    # resume
  end

  def ignore
    # resume
    yield
  ensure
    # hold
  end
end
```

And the code becoming:

```
SignalHolder.hold do
  super { SignalHolder.ignore { yield } }
end
```

The above is not even a suggestion it's merely to materialise and compare the kind of structural "consistency".

It could simply instead just need a comment to prevent the mind tripwire or help someone who would refactor this code in the future for some reason.

Reply:
Got it, thanks for explaining!

So, in parts:

* Yes, it's OK if the `yield` raises. Specifically:
    * We can call `_native_resume_signals` many times in a row.
    
      As you suspected, it leads to `pthread_sigmask` which I believe on Linux will lead to the `rt_sigprocmask` system call being called.
      I've included a benchmark in the PR of what the cost of this is, and here's a run I did:

      ```
      ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [x86_64-linux]
      Warming up --------------------------------------
             hold / resume   391.856k i/100ms
      Calculating -------------------------------------
             hold / resume      3.829M (± 1.9%) i/s -     38.402M in  10.033530s
      ```

      ...so I think we're fine if we end up calling it a few more times than strictly needed.
    * The root of the issue in the Ruby VM happens because in the affected APIs Ruby silently interprets an interruption as "end of folder/empty folder". When an exception is being raised, the iteration will stop anyway, so there's no longer a concern of the interruption causing a semantic mismatch (at least that I can see).

I've pushed 259e134092eae57cb2ccbea028adc8e8d0fc7a19 to add this note to the code as in hindsight I agree it was not very clear before (and hopefully this clarifies it)